### ISSUE TITLE
[FEATURE] Engine upgrade enforcement

### ISSUE THREAD

**Issue by c3y1huang** [longhorn/longhorn#5842]
## Is your feature request related to a problem? Please describe (👍 if you like this request)

Currently, there is no enforcement of the engine version in the Longhorn cluster.  And Longhorn needs to ensure compatibility with all engine versions. 

Without upgrade enforcement, we cannot remove the deprecated features in the later version.

## Describe the solution you'd like

- Enforce upgrading un-upgraded engines before upgrading Longhorn to the new version.

## Describe alternatives you've considered

`None`

## Additional context

https://github.com/longhorn/longhorn/pull/5457#discussion_r1177364728

**Comment by innobead** [comment]
@c3y1huang We need to create an LEP as well, especially record the change from the current behavior.

**Comment by c3y1huang** [comment]
> @c3y1huang We need to create an LEP as well, especially record the change from the current behavior.

The LEP has already been merged. https://github.com/longhorn/longhorn/pull/6519.

**Comment by yangchiu** [comment]
@innobead Do the `revert to the previous Longhorn version` in https://longhorn.io/docs/1.6.0/deploy/important-notes/#engine-upgrade-enforcement and `downgraded Longhorn to v1.5.x` in https://github.com/longhorn/longhorn-tests/pull/1505/files mean we have to rollback Longhorn? 

But don't we already forbid users from doing that: https://github.com/longhorn/longhorn/issues/5131 https://github.com/longhorn/longhorn/issues/6264?

**Comment by innobead** [comment]
This should mean the upgrade failure, so users need to revert (clean up) the currently installed version before the upgrade failure.

https://github.com/longhorn/longhorn/issues/5131 means, that after upgrading successfully, we don't allow users to revert. 

If the document is unclear, we need to adjust the wording. cc @c3y1huang

**Comment by yangchiu** [comment]
> This means the upgrade failure, so users need to revert to the currently installed version before the upgrade failure.

But in https://github.com/longhorn/longhorn/issues/6264, didn't he also encounter upgrade failure but we still didn't expect he to revert?

**Comment by innobead** [comment]
> > This means the upgrade failure, so users need to revert to the currently installed version before the upgrade failure.
> 
> But in #6264, didn't he also encounter upgrade failure but we still didn't expect he to revert?

Users can revert back if an upgrade fails, except for successful upgrades.

**Comment by innobead** [comment]
As discussed with @yangchiu , the context has been made clear. 

- If the upgrade fails, users can `recover` to the original version, since longhorn-manager DS has failed.
- If the upgrade succeeds, users can't `rollback` to the previous version after 1.6.0. For 1.5 -> 1.4, this is allowed, as the upgrade enforcement was introduced in 1.5.   

P.S. Since the performance will take effect after users upgrade to 1.6, we need to have upgrade performance testing in our release testing. cc @yangchiu @khushboo-rancher 

cc @mantissahz

**Comment by yangchiu** [comment]
Verified passed on master-head (longhorn-manager b517601) following [test plan](https://github.com/longhorn/longhorn-tests/pull/1505). When there is a volume using an incompatible engine image, longhorn manager get stuck in `CrashLoopBackOff` during upgrade and error message shown:
```
time="2023-08-17T03:03:20Z" level=fatal msg="Error starting manager: failed checking Engine upgarde path: incompatible Engine ei-7fa7c208 client API version: found version 7 is below required minimal version 8"
```
If install Longhorn by helm, if there is a volume using an incompatible engine image, upgrade will fail with error message:
```
$ helm upgrade longhorn ./longhorn/chart/ -n longhorn-system --debug
WARNING: Kubernetes configuration file is group-readable. This is insecure. Location: /home/yang/Code/automation/terraform2/aws/ubuntu/k3s.yaml
WARNING: Kubernetes configuration file is world-readable. This is insecure. Location: /home/yang/Code/automation/terraform2/aws/ubuntu/k3s.yaml
upgrade.go:150: [debug] preparing upgrade for longhorn
upgrade.go:158: [debug] performing update for longhorn
upgrade.go:330: [debug] creating upgraded release for longhorn
client.go:478: [debug] Starting delete for "longhorn-pre-upgrade" Job
client.go:482: [debug] Ignoring delete failure for "longhorn-pre-upgrade" batch/v1, Kind=Job: jobs.batch "longhorn-pre-upgrade" not found
client.go:134: [debug] creating 1 resource(s)
client.go:706: [debug] Watching for changes to Job longhorn-pre-upgrade with timeout of 5m0s
client.go:734: [debug] Add/Modify event for longhorn-pre-upgrade: ADDED
client.go:773: [debug] longhorn-pre-upgrade: Jobs active: 1, jobs failed: 0, jobs succeeded: 0
client.go:734: [debug] Add/Modify event for longhorn-pre-upgrade: MODIFIED
client.go:773: [debug] longhorn-pre-upgrade: Jobs active: 0, jobs failed: 0, jobs succeeded: 0
client.go:734: [debug] Add/Modify event for longhorn-pre-upgrade: MODIFIED
client.go:478: [debug] Starting delete for "longhorn-pre-upgrade" Job
upgrade.go:442: [debug] warning: Upgrade "longhorn" failed: pre-upgrade hooks failed: 1 error occurred:
	* job failed: BackoffLimitExceeded


Error: UPGRADE FAILED: pre-upgrade hooks failed: 1 error occurred:
	* job failed: BackoffLimitExceeded


helm.go:84: [debug] pre-upgrade hooks failed: 1 error occurred:
	* job failed: BackoffLimitExceeded


UPGRADE FAILED
main.newUpgradeCmd.func2
	helm.sh/helm/v3/cmd/helm/upgrade.go:209
github.com/spf13/cobra.(*Command).execute
	github.com/spf13/cobra@v1.6.1/command.go:916
github.com/spf13/cobra.(*Command).ExecuteC
	github.com/spf13/cobra@v1.6.1/command.go:1044
github.com/spf13/cobra.(*Command).Execute
	github.com/spf13/cobra@v1.6.1/command.go:968
main.main
	helm.sh/helm/v3/cmd/helm/helm.go:83
runtime.main
	runtime/proc.go:250
runtime.goexit
	runtime/asm_amd64.s:1598
```
Once upgrade the incompatible engine image to a compatible one, and upgrade Longhorn again, the above error messages gone and everything works well.

**Comment by innobead** [comment]
Have we tested non-upgrade cases? For example, users will fail to update the engine image of a volume when using an incompatible image? or it has been covered in the e2e? I believe @c3y1huang could have those test cases, PRs.

**Comment by c3y1huang** [comment]
> Have we tested non-upgrade cases? For example, users will fail to update the engine image of a volume when using an incompatible image? or it has been covered in the e2e? I believe @c3y1huang could have those test cases, PRs.

https://github.com/longhorn/longhorn-tests/blob/master/docs/content/manual/release-specific/v1.6.0/test-engine-version-enforcement.md?plain=1#L41-L42

**Comment by innobead** [comment]
> > Have we tested non-upgrade cases? For example, users will fail to update the engine image of a volume when using an incompatible image? or it has been covered in the e2e? I believe @c3y1huang could have those test cases, PRs.
> 
> [longhorn/longhorn-tests@`master`/docs/content/manual/release-specific/v1.6.0/test-engine-version-enforcement.md?plain=1#L41-L42](https://github.com/longhorn/longhorn-tests/blob/master/docs/content/manual/release-specific/v1.6.0/test-engine-version-enforcement.md?plain=1&rgh-link-date=2023-12-28T04%3A27%3A36Z#L41-L42)

@yangchiu Can you check :point_up:

**Comment by c3y1huang** [comment]
Reopen for clarification on the tested cases.

**Comment by yangchiu** [comment]
> > > Have we tested non-upgrade cases? For example, users will fail to update the engine image of a volume when using an incompatible image? or it has been covered in the e2e? I believe @c3y1huang could have those test cases, PRs.
> > 
> > 
> > [longhorn/longhorn-tests@`master`/docs/content/manual/release-specific/v1.6.0/test-engine-version-enforcement.md?plain=1#L41-L42](https://github.com/longhorn/longhorn-tests/blob/master/docs/content/manual/release-specific/v1.6.0/test-engine-version-enforcement.md?plain=1&rgh-link-date=2023-12-28T04%3A27%3A36Z#L41-L42)
> 
> @yangchiu Can you check ☝️

Verified passed on master-head (longhorn-manager b517601) following [test plan](https://github.com/longhorn/longhorn-tests/pull/1505).

**Comment by innobead** [comment]
@c3y1huang the current design is at least one version support for existing volumes before the upgrade, right? for example, if volumes are using 1.5 images, they are allowed to upgrade to 1.6, but if less than 1.5, users need to upgrade the engine image version before upgrade, right?

**Comment by innobead** [comment]
This needs to be re-tested after https://github.com/longhorn/longhorn/issues/7539 is merged. 

cc @longhorn/qa

**Comment by yangchiu** [comment]
Tested on master-head (longhorn/longhorn 11e1a16) following [the test plan](https://github.com/longhorn/longhorn-tests/pull/1505).

After upgrade to master-head, and check the engine image status. `master-head` engine image `INCOMPATIBLE` is `false`, `v1.4.x-head` engine image `INCOMPATIBLE` is `true`, but `v1.5.3` engine image `INCOMPATIBLE` is `empty`:
```
NAME          INCOMPATIBLE   STATE      IMAGE                                    REFCOUNT   BUILDDATE   AGE
ei-68f17757                  deployed   longhornio/longhorn-engine:v1.5.3        10         60d         9m48s
ei-4d51cf83   true           deployed   longhornio/longhorn-engine:v1.4.x-head   0          21d         12m
ei-b907910b   false          deployed   longhornio/longhorn-engine:master-head   0          5d17h       2m
```

Also on the UI, there is no way to tell if the engine image is incompatible, all engine images just show `deployed` status:
![ei](https://github.com/longhorn/longhorn/assets/5709208/76d9b0fd-17f2-4be6-97b7-370afb23b163)

Not sure if it's expected behavior.

cc @c3y1huang

**Comment by c3y1huang** [comment]
> Tested on master-head (longhorn/longhorn [11e1a16](https://github.com/longhorn/longhorn/commit/11e1a1659c4287bbbd570c190b19822a7dad0279)) following [the test plan](https://github.com/longhorn/longhorn-tests/pull/1505).
> 
> After upgrade to master-head, and check the engine image status. `master-head` engine image `INCOMPATIBLE` is `false`, `v1.4.x-head` engine image `INCOMPATIBLE` is `true`, but `v1.5.3` engine image `INCOMPATIBLE` is `empty`:
> 
> ```
> NAME          INCOMPATIBLE   STATE      IMAGE                                    REFCOUNT   BUILDDATE   AGE
> ei-68f17757                  deployed   longhornio/longhorn-engine:v1.5.3        10         60d         9m48s
> ei-4d51cf83   true           deployed   longhornio/longhorn-engine:v1.4.x-head   0          21d         12m
> ei-b907910b   false          deployed   longhornio/longhorn-engine:master-head   0          5d17h       2m
> ```
> 
> Also on the UI, there is no way to tell if the engine image is incompatible, all engine images just show `deployed` status: ![ei](https://private-user-images.githubusercontent.com/5709208/296921312-76d9b0fd-17f2-4be6-97b7-370afb23b163.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUzNzUzOTYsIm5iZiI6MTcwNTM3NTA5NiwicGF0aCI6Ii81NzA5MjA4LzI5NjkyMTMxMi03NmQ5YjBmZC0xN2YyLTRiZTYtOTdiNy0zNzBhZmIyM2IxNjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDExNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMTZUMDMxODE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2I4ODhhYmFhNTAzNTRjODlhOTNmZGIzN2QwOGY0NzMxNzZlN2I5NTI0ZTlhYzIzYWIwYjI0MzAxMmE0NWU2MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ZF5DKy0pCN6vz9uR1oCBNLJ-VhgRVdrEXHcRQQzt6uA)
> 
> Not sure if it's expected behavior.
> 
> cc @c3y1huang

cc @mantissahz

**Comment by mantissahz** [comment]
After this PR https://github.com/longhorn/longhorn/issues/7683 is merged, it will show v1.4.4 engine image `Incompatible` on the Status column.

But v1.5.3 engine image `INCOMPATIBLE` should not be empty.

**Comment by yangchiu** [comment]
Verified passed on master-head (longhorn-manager aaeae7e) and v1.6.x-head (longhorn-manager 43919ce) following [test plan](https://github.com/longhorn/longhorn-tests/blob/3eb1dfaa3723f74e2421a222589769231d90f301/docs/content/manual/release-specific/v1.6.0/test-engine-version-enforcement.md).

And issues in https://github.com/longhorn/longhorn/issues/5842#issuecomment-1893007782 are no longer present:

![imcomp](https://github.com/longhorn/longhorn/assets/5709208/37b06a10-44dd-46e3-8571-8d03f1dbf3d3)

```
$ kubectl get lhei -n longhorn-system
NAME          INCOMPATIBLE   STATE      IMAGE                                    REFCOUNT   BUILDDATE   AGE
ei-a786fad2   false          deployed   longhornio/longhorn-engine:v1.5.x-head   10         28d         8m15s
ei-4d51cf83   true           deployed   longhornio/longhorn-engine:v1.4.x-head   0          23d         11m
ei-b907910b   false          deployed   longhornio/longhorn-engine:master-head   0          97m         93s
```
